core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -673,10 +666,42 @@ namespace kagome::blockchain {
   }
 
   outcome::result<void> BlockTreeImpl::markAsRevertedBlocks(
-      const std::vector<primitives::BlockInfo> &blocks) {
-    // TODO Must be implemented
-#warning "Must be implemented"
-    return outcome::success();
+      const std::vector<primitives::BlockHash> &block_hashes) {
+    return block_tree_data_.exclusiveAccess(
+        [&](auto &p) -> outcome::result<void> {
+          bool need_to_refresh_best = false;
+          auto best = bestBlockNoLock(p);
+          for (const auto &block_hash : block_hashes) {
+            auto tree_node = p.tree_->getRoot().findByHash(block_hash);
+            if (tree_node == nullptr) {
+              SL_WARN(
+                  log_, "Block {} doesn't exists in block tree", block_hash);
+              continue;
+            }
+
+            if (not tree_node->reverted) {
+              std::multiset<std::shared_ptr<TreeNode>> to_revert{
+                  std::move(tree_node)};
+              while (not to_revert.empty()) {
+                auto node = to_revert.extract(to_revert.begin());
+                BOOST_ASSERT(not node.empty());
+
+                auto &reverting_tree_node = node.value();
+                reverting_tree_node->reverted = true;
```

Comment:
[Rust distinguishes](https://github.com/paritytech/polkadot-sdk/blob/771c3fbde7ba3849298d6fc7df390fc6b5a04d2f/polkadot/node/core/chain-selection/src/lib.rs#L77-L84) explicit (disputed, [POL1 digest](https://github.com/paritytech/polkadot-sdk/blob/771c3fbde7ba3849298d6fc7df390fc6b5a04d2f/polkadot/node/core/chain-selection/src/lib.rs#L623)) and inherited revert.

Rare incidents, when force recovery is used may require node restart.
Maybe add comment.

Reply:
Can you provide more details for understanding?